platform/kube/controller_test.go

```diff
@@ -466,67 +467,97 @@ func makeService(n, ns string, cl kubernetes.Interface, t *testing.T) {
-	createPod(clientSet, map[string]string{"app": "prod-app"}, "pod2", "nsA", "acct2", t)
-	createPod(clientSet, map[string]string{"app": "prod-app"}, "pod3", "nsA", "acct3", t)
-	createPod(clientSet, map[string]string{"app": "prod-app"}, "pod4", "nsA", "acct3", t)
-	createPod(clientSet, map[string]string{"app": "prod-app"}, "pod5", "nsB", "acct4", t)
-
 	controller := NewController(&Client{client: clientSet}, ControllerConfig{
 		Namespace:    "default",
 		ResyncPeriod: resync,
 	})
 
-	createService(controller, "svc1", "nsA", map[string]string{"app": "prod-app"}, t)
-	createService(controller, "svc2", "nsA", map[string]string{"app": "staging-app"}, t)
+	createPod(controller, map[string]string{"app": "test-app"}, "pod1", "nsA", "acct1", t)
+	createPod(controller, map[string]string{"app": "prod-app"}, "pod2", "nsA", "acct2", t)
+	createPod(controller, map[string]string{"app": "prod-app"}, "pod3", "nsA", "acct3", t)
+	createPod(controller, map[string]string{"app": "prod-app"}, "pod4", "nsA", "acct3", t)
+	createPod(controller, map[string]string{"app": "prod-app"}, "pod5", "nsB", "acct4", t)
+
+	// Populate pod cache.
+	controller.pods.keys["128.0.0.1"] = "nsA/pod1"
+	controller.pods.keys["128.0.0.2"] = "nsA/pod2"
+	controller.pods.keys["128.0.0.3"] = "nsA/pod3"
+	controller.pods.keys["128.0.0.4"] = "nsA/pod4"
+	controller.pods.keys["128.0.0.5"] = "nsB/pod5"
+
+	createService(controller, "svc1", "nsA", []int32{8080}, map[string]string{"app": "prod-app"}, t)
+	createService(controller, "svc2", "nsA", []int32{8081}, map[string]string{"app": "staging-app"}, t)
+
+	svc1Ips := []string{"128.0.0.1", "128.0.0.2"}
+	portNames := []string{"test-port"}
+	createEndpoints(controller, "svc1", "nsA", portNames, svc1Ips, t)
 
 	hostname := serviceHostname("svc1", "nsA")
-	sa, err := controller.GetIstioServiceAccounts(hostname)
-	if err != nil {
-		t.Error("Error returned: ", err)
-	} else if len(sa) != 2 ||
-		!(sa[0] == "istio:acct2.nsA.cluster.local" && sa[1] == "istio:acct3.nsA.cluster.local" ||
-			sa[0] == "istio:acct3.nsA.cluster.local" && sa[1] == "istio:acct2.nsA.cluster.local") {
-		t.Error("Failure: The resolved service accounts are not correct: ", sa)
+	sa := controller.GetIstioServiceAccounts(hostname, []string{"test-port"})
+	sort.Sort(sort.StringSlice(sa))
+	expected := []string{"istio:acct1.nsA.cluster.local", "istio:acct2.nsA.cluster.local"}
+	if !reflect.DeepEqual(sa, expected) {
+		t.Errorf("Unexpected service accounts %v (expecting %v)", sa, expected)
 	}
 
 	hostname = serviceHostname("svc2", "nsA")
-	sa, err = controller.GetIstioServiceAccounts(hostname)
-	if err != nil {
-		t.Error("Error returned: ", err)
-	} else if len(sa) != 0 {
+	sa = controller.GetIstioServiceAccounts(hostname, []string{})
+	if len(sa) != 0 {
 		t.Error("Failure: Expected to resolve 0 service accounts, but got: ", sa)
 	}
+}
+
+func createEndpoints(controller *Controller, name, namespace string, portNames, ips []string, t *testing.T) {
```

Comment:
nit: I think we should switch to using "real" kubernetes interface instead of the fake one. Creation of the endpoints is really the job of k8s controllers.